src/main/java/com/marklogic/spark/reader/file/ArchiveFileReader.java

```diff
@@ -0,0 +1,75 @@
+package com.marklogic.spark.reader.file;
+
+import com.marklogic.spark.ConnectorException;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.connector.read.PartitionReader;
+import org.apache.spark.unsafe.types.ByteArray;
+import org.apache.spark.unsafe.types.UTF8String;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+public class ArchiveFileReader implements PartitionReader<InternalRow> {
+    private static final Logger logger = LoggerFactory.getLogger(ArchiveFileReader.class);
+    private final String path;
+    private final ZipInputStream zipInputStream;
+    private ZipEntry currentZipEntry;
+  //  private ZipEntry nextZipEntry;
+    int count;
+    byte[] content;
+
+    public ArchiveFileReader(FilePartition partition, FileContext fileContext) {
+        this.path = partition.getPath();
+        this.zipInputStream = new ZipInputStream(fileContext.open(partition));
+        count = 0;
+    }
+
+    @Override
+    public boolean next() throws IOException {
+        currentZipEntry = FileUtil.findNextFileEntry(zipInputStream);
+        return currentZipEntry != null;
+    }
+
+    @Override
+    public InternalRow get() {
+
+        String zipEntryName = currentZipEntry.getName();
+        if (logger.isTraceEnabled()) {
+            logger.trace("Reading zip entry {} from zip file {}.", zipEntryName, this.path);
+        }
+        String uri = zipEntryName.startsWith("/") ?
+            this.path + zipEntryName :
+            this.path + "/" + zipEntryName;
+        if(uri.contains(".metadata"))
+            uri = uri.substring(0, uri.indexOf(".metadata"));
+        content = readZipEntry(count);
+
+        long length = content.length;
+        return new GenericInternalRow(new Object[]{
+            UTF8String.fromString(uri), null, length, ByteArray.concat(content)
+        });
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.zipInputStream.close();
+    }
+
+    private byte[] readZipEntry(int count) {
+        try {
+            count++;
+            byte[] content = FileUtil.readBytes(zipInputStream);
+            if(count%2 != 0 && next()){
+                return ByteArray.concat(content, FileUtil.readBytes(zipInputStream));
```

Comment:
We don't want to concat the next entry. Instead, for an NT archive file, each time we read a zip entry, we know that the next entry will be a metadata file. So we want pseudocode like this:

```
byte[] content = readZipEntry
metadataEntry = findNextFileEntry
// Throw error if this doesn't end with ".metadata", which means it's not a valid archive file. 
byte[] metadata = readZipEntry
// Make an InternalRow using the content data and the metadata.
```

You can use `DocumentMetadataHandle.fromBuffer(byte[])` to convert the byte array into a `DocumentMetadataHandle` instance.

Reply:
I just started the story for reading an MLCP archive, and I think the approach I'm taking in there should be the same as what's needed here. Will share soon.